pkg/util/pki/kube.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	experimentalapi "github.com/jetstack/cert-manager/pkg/apis/experimental/v1alpha1"
+)
+
+// GenerateTemplate will create a x509.Certificate for the given
+// CertificateSigningRequest resource
+func GenerateTemplateFromCertificateSigningRequest(csr *certificatesv1.CertificateSigningRequest) (*x509.Certificate, error) {
+	duration := cmapi.DefaultCertificateDuration
+	requestedDuration, ok := csr.Annotations[experimentalapi.CertificateSigningRequestDurationAnnotationKey]
+	if ok {
+		dur, err := time.ParseDuration(requestedDuration)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse requested duration on annotation %q: %w",
+				experimentalapi.CertificateSigningRequestDurationAnnotationKey, err)
+		}
+		duration = dur
+	}
+	fmt.Println(duration)
+
+	ku, eku, err := BuildKeyUsagesKube(csr.Spec.Usages)
+	if err != nil {
+		return nil, err
+	}
+
+	isCA := csr.Annotations[experimentalapi.CertificateSigningRequestIsCAAnnotationKey] == "true"
+
+	return GenerateTemplateFromCSRPEMWithUsages(csr.Spec.Request, duration, isCA, ku, eku)
+}
+
+func BuildKeyUsagesKube(usages []certificatesv1.KeyUsage) (ku x509.KeyUsage, eku []x509.ExtKeyUsage, err error) {
+	var unk []certificatesv1.KeyUsage
+	if len(usages) == 0 {
+		usages = []certificatesv1.KeyUsage{certificatesv1.UsageDigitalSignature, certificatesv1.UsageKeyEncipherment}
```

Comment:
suggestion: should these default values come from https://github.com/jetstack/cert-manager/blob/a80198c03d108e4ceb7e5aceb0f2fd78be5db39d/pkg/apis/certmanager/v1/types.go#L203?

(I recognise that will require converting a `certmangerv1.KeyUsage` to a `certificatesv1.KeyUsage`, but I think it'd be good to re-use that function)

Reply:
Not sure I understand the intention with reusing this function. We are using a separate API group and are different types.

Reply:
"Default cert-manager key usages for certificates" is a useful concept and if we're going to have it as a concept we might as well have one place to change it rather than multiple places where we can forget one of them. I suppose we'd have, say, in pseudocode:

```go
func DefaultKeyUsages() []x509.KeyUsage
```

in pkg/internal/ somewhere. I don't think it's required for this PR to be merged.

Reply:
Let's leave this for now. We can always add the Default to the API at a later date.